MetadataExtractor/IO/RandomAccessStream.cs

```diff
@@ -0,0 +1,561 @@
+﻿// Copyright (c) Drew Noakes and contributors. All Rights Reserved. Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using JetBrains.Annotations;
+
+namespace MetadataExtractor.IO
+{
+    /// <summary>Reads and buffers data in chunks and provides methods for reading data types</summary>
+    /// <remarks>
+    /// This class implements buffered reading of data typically for use with <see cref="ReaderInfo"/> 
+    /// objects and provides methods for reading data types from it. Data is captured in configurable 
+    /// chunks for efficiency. Both seekable and non-seekable streams are supported.
+    /// </remarks>
+    /// <author>Kevin Mott https://github.com/kwhopper</author>
+    /// <author>Drew Noakes https://drewnoakes.com</author>
+    public class RandomAccessStream
+    {
+        private Stream? p_inputStream;
+        private long p_streamLength = -1;
+
+        //private readonly List<ReaderInfo> rdrList = new List<ReaderInfo>();
+        private bool p_isStreamFinished;
+
+        private const int DefaultChunkLength = 4 * 1024;
+        private readonly int p_chunkLength;
+        public Dictionary<long, byte[]> p_chunks = new Dictionary<long, byte[]>();
+
+        public RandomAccessStream(Stream stream, long streamLength = -1)
+        {
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
+
+            if (streamLength == -1)
+                streamLength = stream.CanSeek ? stream.Length : int.MaxValue;
+
+            p_inputStream = stream;
+            CanSeek = stream.CanSeek;
+
+            p_chunkLength = DefaultChunkLength;
+            p_streamLength = streamLength;
+        }
+
+        public RandomAccessStream(byte[] bytes)
+        {
+            if (bytes == null)
+                throw new ArgumentNullException(nameof(bytes));
+
+            CanSeek = true;
+
+            // Setting these values makes p_inputStream irrelevant
+            // TODO: break the byte array up into DefaultChunkLength chunks
```

Comment:
Why break it up?

Reply:
If the array is large (like an uncompressed ICC byte array) one thought was to break it up in the underlying buffer to avoid holding two LOH objects. It doesn't avoid making that large array in the first place.

This is sort of a reminder that these classes are only a beginning. It's potentially possible to avoid byte arrays entirely for some problem types like large text or XML and immediately cache to disk (where possible), which could then naturally work as a new RandomAccessStream.